include/sycl/algorithm/equal.hpp

```diff
@@ -0,0 +1,156 @@
+  MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+
+*/
+
+#ifndef __SYCL_IMPL_ALGORITHM_EQUAL__
+#define __SYCL_IMPL_ALGORITHM_EQUAL__
+
+#include <algorithm>
+#include <functional>
+#include <iostream>
+#include <type_traits>
+
+// SYCL helpers header
+#include <sycl/algorithm/algorithm_composite_patterns.hpp>
+#include <sycl/algorithm/buffer_algorithms.hpp>
+#include <sycl/helpers/sycl_buffers.hpp>
+#include <sycl/helpers/sycl_differences.hpp>
+
+namespace sycl {
+namespace impl {
+
+#ifdef SYCL_PSTL_USE_OLD_ALGO
+
+/* equal.
+* @brief Implementation of the command group that submits an equal kernel.
+*/
+template <class ExecutionPolicy, class ForwardIt1, class ForwardIt2,
+          class BinaryPredicate>
+bool equal(ExecutionPolicy& exec, ForwardIt1 first1, ForwardIt1 last1,
+           ForwardIt2 first2, ForwardIt2 last2, BinaryPredicate p) {
+  cl::sycl::queue q(exec.get_queue());
+
+  auto size1 = sycl::helpers::distance(first1, last1);
+  auto size2 = sycl::helpers::distance(first2, last2);
+
+  if (size1 != size2) {
+    return false;
+  }
+
+  if (size1 < 1) {
+    return true;
+  }
+
+  auto device = q.get_device();
+
+  auto local_size = std::min(
+      device.get_info<cl::sycl::info::device::max_work_group_size>(), size1);
+  auto global_size = exec.calculateGlobalSize(size1, local_size);
+
+  auto buf1 = sycl::helpers::make_const_buffer(first1, last1);
+  auto buf2 = sycl::helpers::make_const_buffer(first2, last2);
+  auto bufR = cl::sycl::buffer<bool, 1>(cl::sycl::range<1>(size1));
+
+  do {
+    int passes = 0;
+
+    auto f = [global_size, local_size, passes, &buf1, &buf2, &bufR,
+              p](cl::sycl::handler& h) mutable {
+      cl::sycl::nd_range<1> r{
+          cl::sycl::range<1>{std::max(global_size, local_size)},
+          cl::sycl::range<1>{local_size}};
+      auto a1 = buf1.template get_access<cl::sycl::access::mode::read>(h);
+      auto a2 = buf2.template get_access<cl::sycl::access::mode::read>(h);
+      auto aR = bufR.template get_access<cl::sycl::access::mode::read_write>(h);
+      cl::sycl::accessor<bool, 1, cl::sycl::access::mode::read_write,
+                         cl::sycl::access::target::local>
+          scratch(cl::sycl::range<1>(local_size), h);
+
+      h.parallel_for<typename ExecutionPolicy::kernelName>(
+          r, [a1, a2, aR, scratch, local_size, global_size, passes,
+              p](cl::sycl::nd_item<1> id) {
+            auto r =
+                ReductionStrategy<bool>(local_size, global_size, id, scratch);
+            if (passes == 0) {
+              r.workitem_get_from(p, a1, a2);
+            } else {
+              r.workitem_get_from(aR);
+            }
+            r.combine_threads(std::logical_and<bool>{});
+            r.workgroup_write_to(aR);
+          });  // end kernel
+    };         // end command group
+
+    q.submit(f);
+    global_size = global_size / local_size;
+    ++passes;
+  } while (global_size > 1);
```

Comment:
Instead of doing the full reduction, maybe we can "short-circuit" the evaluation.
An additional buffer of size one, initially set to false, can be set to true inside the kernel whenever the predicate is false.
We can copy back to the host this single element at each iteration and check if its true, in which case, all_of is false and we don't need to continue execution.

Reply:
That is undefined behaviour in OpenCL C 2.0 and C++ without synchronisation. I have had failures when threads write the exact same value to the same location without synchronisation on Intel CPUs.

Reply:
You are correct. Not sure if it will be worse in performance an atomic or the extra kernel(s) launch(s). Ignore my comment for now. Let's keep the algorithms simple: we can always produce optimized versions later.